### DISCUSSION TITLE
Raise an event from runtime engine when a new module/script/function is added to cache.

### DISCUSSION THREAD

**Discussion by HaloX69** [PowerShell/PowerShell#15122] [Q&A]
I am looking to enable specific logging as well as override specific methods on objects when certain modules/classes/scripts/functions are loaded.  I would have a base script that other consuming scripts use as a base helper.  This would require a script/class that runs and when it detects changes to the scriptblock case say just for scriptblocks, which would address 90% of the needs, the problem is the ConcurrentDictionary object doesn't raise any events.  I was working with creating an observable ConcurrentDictionary, but then there we issues with casting (I would obviously need to create a custom extension method) and then making the backend engine leverage this object instead.

I looked at runspaces, but native events are limited, and still any changes to the scriptblock cache is not propagated from the host runspace to the new runspace.

This seems like a common need for advance PowerShell solutions.  Any suggestions?

**Comment by SeeminglyScience** [comment]
There's an open issue for a module loading event (#13493).

I can see it being useful as a general `CommandLoaded` event, but at the individual scriptblock level I'm not sure.  Can you elaborate a bit on your use cases?

**Comment by HaloX69** [comment]
What if I want to inject code into a scriptblock when added to the cache?  I have a function that enumerates all input parameters default/provided values and logs them. I want to inject into each function the call to this function if verbose is enabled. It is tedious to have to provide this simple code block in each script/function.

The module is used by many dev groups and they use other modules/scripts that they don’t control the source so being able to know and access a scriptblock would be very useful.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
